RepaintRects need to be recomputed when state changes that affects them

smfr opened the pull request:
#### abd0e960c37bcf3ba63944e9544bffb45be80722
<pre>
RepaintRects need to be recomputed when state changes that affects them
<a href="https://bugs.webkit.org/show_bug.cgi?id=284351">https://bugs.webkit.org/show_bug.cgi?id=284351</a>
<a href="https://rdar.apple.com/141198097">rdar://141198097</a>

Reviewed by Alan Baradlay.

With the PR in bug 267252 applied, these two tests hit assertions in repaint rect
verification (RenderLayer.cpp:1230):

imported/blink/fast/hidpi/border-background-align.html
printing/iframe-print.html

This is because bug 267252 eliminates a layout that happens when the state of `isPrinting`
or deviceScaleFactor changes, so `updateLayerPositions()` doesn&apos;t do a full traversal, and
we don&apos;t recompute new repaint rects.

printing/iframe-print.html asserts because the return value of `RenderView::computeVisibleRectsInContainer()`
is affecting by printing.
imported/blink/fast/hidpi/border-background-align.html asserts because `deviceScaleFactor` affects rect rounding.

Fix by tracking these in `RepaintRectEnvironment` in LocalFrameViewLayoutContext, and when they change, forcing
a full RenderLayer traversal via a new flag added to `invalidationLayerPositionsFlags()`.

* Source/WebCore/page/LocalFrameViewLayoutContext.cpp:
(WebCore::LocalFrameViewLayoutContext::flushUpdateLayerPositions):
* Source/WebCore/page/LocalFrameViewLayoutContext.h:
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::RenderLayer):
(WebCore::RenderLayer::updateLayerPositionsAfterLayout):
* Source/WebCore/rendering/RenderLayer.h:
* Source/WebCore/rendering/RenderObject.h: Fix typo

Canonical link: <a href="https://commits.webkit.org/287607@main">https://commits.webkit.org/287607@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/71faef33592197575d35730c52cf6b325a05f67b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/80216 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/59218 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/33629 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/84733 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/31192 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/82327 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/68279 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/7517 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/62705 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/20518 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/83285 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/52775 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/73052 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/43013 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/50104 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/27207 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/29653 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/71233 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/27717 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/86166 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/7436 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/5264 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/70979 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/7611 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/68892 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/70221 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17493 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/14215 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/13163 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/7400 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/12917 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/7239 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/10759 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/9044 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on Source/WebCore/page/LocalFrameViewLayoutContext.cpp:
```diff
@@ -344,12 +344,19 @@ void LocalFrameViewLayoutContext::flushUpdateLayerPositions()
         return;
 
+    CheckedPtr view = renderView();
+    if (!view)
+        return;
```
nit. I usually add an ASSERT_NOT_REACHED if it is an unexpected state.

alanbaradlay commented on Source/WebCore/rendering/RenderLayer.h:
```diff
@@ -1032,6 +1032,7 @@ class RenderLayer final : public CanMakeSingleThreadWeakPtr<RenderLayer>, public
         Seen3DTransformedLayer              = 1 << 7,
         SeenCompositedScrollingLayer        = 1 << 8,
         SubtreeNeedsUpdate                  = 1 << 9,
+        EnvironmentChanged                  = 1 << 10,
```
I think this needs some comment. it's not super obvious what "environment" here means.

smfr committed:
RepaintRects need to be recomputed when state changes that affects them
https://bugs.webkit.org/show_bug.cgi?id=284351
rdar://141198097

Reviewed by Alan Baradlay.

With the PR in bug 267252 applied, these two tests hit assertions in repaint rect
verification (RenderLayer.cpp:1230):

imported/blink/fast/hidpi/border-background-align.html
printing/iframe-print.html

This is because bug 267252 eliminates a layout that happens when the state of `isPrinting`
or deviceScaleFactor changes, so `updateLayerPositions()` doesn't do a full traversal, and
we don't recompute new repaint rects.

printing/iframe-print.html asserts because the return value of `RenderView::computeVisibleRectsInContainer()`
is affecting by printing.
imported/blink/fast/hidpi/border-background-align.html asserts because `deviceScaleFactor` affects rect rounding.

Fix by tracking these in `RepaintRectEnvironment` in LocalFrameViewLayoutContext, and when they change, forcing
a full RenderLayer traversal via a new flag added to `invalidationLayerPositionsFlags()`.

* Source/WebCore/page/LocalFrameViewLayoutContext.cpp:
(WebCore::LocalFrameViewLayoutContext::flushUpdateLayerPositions):
* Source/WebCore/page/LocalFrameViewLayoutContext.h:
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::RenderLayer):
(WebCore::RenderLayer::updateLayerPositionsAfterLayout):
* Source/WebCore/rendering/RenderLayer.h:
* Source/WebCore/rendering/RenderObject.h: Fix typo

Canonical link: https://commits.webkit.org/287607@main

mattwoodrow commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -1077,14 +1080,18 @@ void RenderLayer::updateLayerPositionsAfterLayout(RenderElement::LayoutIdentifie
 #endif
+
+    if (environmentChanged)
+        setSelfAndDescendantsNeedPositionUpdate();
```
We shouldn't need to do both of setting the flag and marking the layer. Either are sufficient to trigger recomputation of layer positions of the subtree.

The general idea is that the 'need position update' bits (self, self+children, self+alldescendants) are for changes that happen outside of updating layer positions. The flags are for state changes that are detected during the walk, and we newly need to walk descendants.

The flags could probably just be replaced by adding new dirty bits mid-walk, but they have existing meaning too.
